ksqldb-engine/src/main/java/io/confluent/ksql/planner/plan/LogicRewriter.java

```diff
@@ -0,0 +1,337 @@
+ * specific language governing permissions and limitations under the License.
+ */
+
+package io.confluent.ksql.planner.plan;
+
+import io.confluent.ksql.engine.rewrite.ExpressionTreeRewriter;
+import io.confluent.ksql.engine.rewrite.ExpressionTreeRewriter.Context;
+import io.confluent.ksql.execution.expression.tree.BooleanLiteral;
+import io.confluent.ksql.execution.expression.tree.ComparisonExpression;
+import io.confluent.ksql.execution.expression.tree.ComparisonExpression.Type;
+import io.confluent.ksql.execution.expression.tree.Expression;
+import io.confluent.ksql.execution.expression.tree.LogicalBinaryExpression;
+import io.confluent.ksql.execution.expression.tree.NotExpression;
+import io.confluent.ksql.execution.expression.tree.QualifiedColumnReferenceExp;
+import io.confluent.ksql.execution.expression.tree.UnqualifiedColumnReferenceExp;
+import io.confluent.ksql.execution.expression.tree.VisitParentExpressionVisitor;
+import io.confluent.ksql.util.KsqlException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+
+public final class LogicRewriter {
+
+
+  private LogicRewriter() {
+  }
+
+  public static Expression rewriteNegations(final Expression expression) {
+    return new ExpressionTreeRewriter<>(new NotPropagator()::process)
+        .rewrite(expression, new NotPropagatorContext());
+  }
+
+  public static Expression rewriteCNF(final Expression expression) {
+    final Expression notPropagated = new ExpressionTreeRewriter<>(new NotPropagator()::process)
+          .rewrite(expression, new NotPropagatorContext());
+    return new ExpressionTreeRewriter<>(new DistributiveLawApplierDOC()::process)
+        .rewrite(notPropagated, null);
+  }
+
+  public static Expression rewriteDNF(final Expression expression) {
+    final Expression notPropagated = new ExpressionTreeRewriter<>(new NotPropagator()::process)
+        .rewrite(expression, new NotPropagatorContext());
+    return new ExpressionTreeRewriter<>(new DistributiveLawApplierCOD()::process)
+        .rewrite(notPropagated, null);
+  }
+
+  public static List<Expression> extractDisjuncts(final Expression expression) {
+    final Expression dnf = rewriteDNF(expression);
+    final DisjunctExtractor disjunctExtractor = new DisjunctExtractor();
+    disjunctExtractor.process(dnf, null);
+    return disjunctExtractor.getDisjuncts();
+  }
+
+  private static final class NotPropagator extends
+      VisitParentExpressionVisitor<Optional<Expression>, Context<NotPropagatorContext>> {
+
+    @Override
+    public Optional<Expression> visitExpression(
+        final Expression node,
+        final Context<NotPropagatorContext> context) {
+      return Optional.empty();
+    }
+
+    @Override
+    public Optional<Expression> visitUnqualifiedColumnReference(
+        final UnqualifiedColumnReferenceExp node,
+        final Context<NotPropagatorContext> context
+    ) {
+      return handlePrimitiveTerm(node, context);
```

Comment:
In this case, my notion of a primitive term is anything that can have a not in front of it, e.g. a boolean column or a boolean literal.  I'm not sure how generic this would be to make a visitor that allowed you to override one method `handlePrimitiveTerm` with this notion of a primitive term baked in.  For example, there are all sorts of literals I don't care about since they make no sense to put a not in front of like `NOT 'foo'`.